Correct semantics for XMLHttpRequest.responseText

philikon opened the pull request:
Accessing the `responseText` property when `responseType` is not `''` or `'text'` should throw. Also, the property is read-only.

**Test Plan:** UIExplorer example, unit tests

facebook-github-bot-3 commented:
By analyzing the blame information on this pull request, we identified @davidaurelio and @grgmo to be potential reviewers.

facebook-github-bot-0 commented:
@philikon updated the pull request.

philikon commented:
I don't know what eslint's problems are. All strings are singlequotes, except for one template string that uses backticks for a good reason (because it contains a bunch of singlequotes inside the string as well as a variable expression.)

satya164 commented:
Can you also add a comment with the link to the spec which describes this behaviour?

davidaurelio commented on Libraries/Network/__tests__/XMLHttpRequestBase-test.js:
```diff
@@ -54,6 +54,25 @@ describe('XMLHttpRequestBase', function(){
 		expect(xhr.readyState).toBe(xhr.OPENED);
     });
 
+    it('should expose responseText correctly', function() {
```
indentation

philikon replied:
Grr, yes. This file is, for some reason, indented with tabs and not spaces. Maybe after this PR I'll fix that because it's super annoying.

davidaurelio commented on Libraries/Network/__tests__/XMLHttpRequestBase-test.js:
```diff
@@ -54,6 +54,25 @@ describe('XMLHttpRequestBase', function(){
     });
 
+    it('should expose responseText correctly', function() {
+		expect(xhr.responseType).toBe('');
```
Please don’t test for this in a test for `responseType`. Either set it, or trust the default (which should be tested in its own test block).

philikon replied:
Ok.

davidaurelio commented:
looks mostly good to me. The spec also demands that the attribute returns `''` if the state is < `LOADING`. That’s working, it’s just not enforced by code or tests.

philikon commented:
> The spec also demands that the attribute returns `''` if the `state is < LOADING`. That’s working, it’s just not enforced by code or tests.

Oh, sure, I can add that. Good catch.

philikon commented:
Sorry, things have been busy. I'll address review comments so we can get this landed.

facebook-github-bot-3 commented:
@philikon updated the pull request.

philikon commented:
Alright @davidaurelio, finally addressed your review comments. Added some more test cases which also revealed a pre-existing bug. Hooray for unit tests!

philikon commented:
Not sure what exactly the circleci failure is, but it seems unrelated. It seems like it's from a WebView test or something.

satya164 commented:
@philikon can you rebase so that the tests run again?

facebook-github-bot-1 commented:
@philikon updated the pull request.

philikon commented:
Yay, tests are green. Can haz merge nao? :)

ide commented:
@facebook-github-bot shipit

facebook-github-bot-5 commented:
Thanks for importing. If you are an FB employee go to [Phabricator](https://our.intern.facebook.com/intern/opensource/github/pull_request/845760482191649/int_phab) to review.

philikon commented on Libraries/Network/__tests__/XMLHttpRequest-test.js:
```diff
@@ -56,15 +56,52 @@ describe('XMLHttpRequest', function(){
 	});
 
-  it('should transition readyState correctly', function() {
+	it('should transition readyState correctly', function() {
```
![spaces vs tabs silicon valley](https://cloud.githubusercontent.com/assets/54918/15690214/69a6d47a-2737-11e6-8f99-90c1927db85f.png)

ide replied:
next thing you'll tell us you used vim over emacs

facebook-github-bot-0 commented:
I tried to merge this pull request into the Facebook internal repo but some checks failed. To unblock yourself please check the following: Does this pull request pass all open source tests on GitHub? If not please fix those. Does the code still apply cleanly on top of GitHub master? If not can please rebase. In all other cases this means some internal test failed, for example a part of a fb app won't work with this pull request. I've added the Import Failed label to this pull request so it is easy for someone at fb to find the pull request and check what failed. If you don't see anyone comment in a few days feel free to comment mentioning one of the core contributors to the project so they get a notification.

philikon commented:
@davidaurelio @nicklockwood halp?

facebook-github-bot-9 commented:
@philikon updated the pull request.

fkgozali commented:
@facebook-github-bot shipit

philikon commented:
Thanks @fkgozali, but it seems like the fb bot doesn't recognize your powers here :)

@ide could I get another import please?

nicklockwood commented:
@facebook-github-bot shipit

philikon commented:
Did we break the bots?

mkonicek commented:
Hmm Kevin Gozali commented in Phabricator:

    FAIL react-native-github/Libraries/Network/__tests__/XMLHttpRequest-test.js (0.32s)
    ● XMLHttpRequest › it should expose responseType correctly

    TypeError: xhr._didReceiveData is not a function

mkonicek commented:
I'm trying again just to see if the bot is working:

@facebook-github-bot shipit

facebook-github-bot-5 commented:
Thanks for importing. If you are an FB employee go to [Phabricator](https://our.intern.facebook.com/intern/opensource/github/pull_request/845760482191649/int_phab) to review.

bestander commented:
Could this fix be responsible for Jest Tests now being skipped, @philikon?

https://circleci.com/gh/facebook/react-native/7709

```
> react-native@1000.0.0 test /home/ubuntu/react-native
> NODE_ENV=test jest "--maxWorkers=1"

Using Jest CLI v12.1.1, jasmine2
...
Running 30 test suites...Warning: The provided value 'arrayblobbuffertextfile' is not a valid 'responseType'.

 PASS  Libraries/Network/__tests__/XMLHttpRequest-test.js (0.393s)
...
Running 1 test suite... PASS  Libraries/WebSocket/__tests__/WebSocket-test.js (0.329s)
13 tests skipped, 453 tests passed (466 total in 50 test suites, run time 26.11s)
```

davidaurelio commented:
@bestander not really:
https://github.com/facebook/react-native/blob/master/Libraries/Network/__tests__/XMLHttpRequest-test.js#L70-L83
https://github.com/facebook/react-native/blob/master/Libraries/Network/XMLHttpRequest.js#L150-L154
